Handle Ctrl+Backspace hotkey for all text boxes

zvirja opened the pull request:
<!-- Please read CONTRIBUTING.md before submitting a pull request -->

Closes #5175

Changes proposed in this pull request:
- Handle `Ctrl+Backspace` hotkey in text inputs for all the dialogs
 
Screenshots before and after (if PR changes UI):
Current:
![current](https://imgur.com/download/oPVuT99)

After:
![expected](https://imgur.com/download/SUayGBs)

What did I do to test the code and ensure quality:
- Performed the manual testing on my environment. Not sure I can create a unit test for that.

Has been tested on (remove any that don't apply):
- GIT 2.10 and above
- Windows 7 and above

gerhardol commented on GitUI/CommandsDialogs/FormCreateBranch.cs:
```diff
@@ -41,6 +41,19 @@ public FormCreateBranch(GitUICommands commands, GitRevision revision)
+        protected override bool ProcessCmdKey(ref Message msg, Keys keyData)
+        {
+            // Handle Ctrl+Backspace combination in the text input (remove last word).
+            // More detail: https://stackoverflow.com/a/1197339/2009373
```
What about the second alternative? 

Should this be applied globally?

zvirja replied:
The global change looks quite dangerous to me (e.g. it could break complex `Message` field in the `Commit` dialog) and should be carefully tested. That's why I decided to proceed with a local change only, where form is tiny as it's easy to test it.

drewnoakes replied:
I use <kbd>Ctrl</kbd>+<kbd>Backspace</kbd> all the time, and it bugs me when a text box doesn't support it. Thanks for this PR.

> Should this be applied globally?

I'd prefer a global fix to a local one.

If it requires overriding `ProcessCmdKey` then it'd be hard to do. I agree the second SO answer looks promising, as it [shows an alternative](https://superuser.com/a/332034/31787).

> The "delete word" behavior was added later, [as an undocumented feature](http://blogs.msdn.com/b/oldnewthing/archive/2007/10/11/5395501.aspx), and only to those Edit controls which use "SHAutoComplete". If a textbox has autocompletion disabled, <kbd>Ctrl</kbd><kbd>Backspace</kbd> will likely not work.

I tried this out in the method we use to walk the control tree to fix DPI issues, and it seems to work well:

```diff

diff --git a/GitExtUtils/GitUI/ControlDpiExtensions.cs b/GitExtUtils/GitUI/ControlDpiExtensions.cs
index af8541452..4d7d5bfdc 100644
--- a/GitExtUtils/GitUI/ControlDpiExtensions.cs
+++ b/GitExtUtils/GitUI/ControlDpiExtensions.cs
@@ -2,6 +2,7 @@
 using System.Collections.Generic;
 using System.ComponentModel;
 using System.Drawing;
+using System.Runtime.InteropServices;
 using System.Windows.Forms;
 using GitExtUtils.GitUI;
 using JetBrains.Annotations;
@@ -95,11 +96,18 @@ public static void AdjustForDpiScaling([NotNull] this Control control)
                         break;
                     }

-                    case TextBoxBase textBox when next.Margin == new Padding(12):
+                    case TextBoxBase textBoxBase when next.Margin == new Padding(12):
                     {
+                        if (textBoxBase is TextBox textBox)
+                        {
+                            const uint SHACF_AUTOSUGGEST_FORCE_OFF = 0x20000000;
+
+                            SHAutoComplete(textBox.Handle, SHACF_AUTOSUGGEST_FORCE_OFF);
+                        }
+
                         // Work around a bug in WinForms where the control's margin gets scaled beyond expectations
                         // see https://github.com/gitextensions/gitextensions/issues/5098
-                        textBox.Margin = DpiUtil.Scale(new Padding(3));
+                        textBoxBase.Margin = DpiUtil.Scale(new Padding(3));
                         EnqueueChildren();
                         break;
                     }
@@ -129,5 +137,8 @@ void EnqueueChildren()
                 }
             }
         }
+
+        [DllImport("shlwapi.dll", ExactSpelling = true, PreserveSig = false)]
+        private static extern int SHAutoComplete(IntPtr hwndEdit, uint flags);
     }
 }
\ No newline at end of file
```

This will only apply to `TextBox`. I ran some tests for a few minutes and it seems to work well across the board.

@zvirja what do you think of the approach outlines in the above diff? It might not apply cleanly to master as I based it off the `v3-ux` branch from #5087.

RussKie replied:
Nice! Let's apply it globally, if possible.

zvirja replied:
@drewnoakes Thanks for the insights! Probably it's indeed better to follow the approach you described.
Let's wait till #5087 is merged and proceed after that!

Could you please ping me when you are ready?

RussKie replied:
We're all sitting and waiting for that moment  🍺 🍿 🕐

drewnoakes replied:
@zvirja #5087 has been merged now so feel free to pick this up again and we'll get it merged in.

zvirja commented:
@drewnoakes @RussKie I found some time and reworked the approach to make the feature global. I'm not sure how Drew googled this way, but it works damn awesome! 😱 

I rewrote the DPI Scaling extension method to the visitor pattern, so I can re-use the tree walking feature we already have. Decided to remove the extension method at all and use the visitor directly. It adds better clarify and look OK to me.

I partially tested it locally and wasn't able to find issues. I worried about commit message dialog, but it works fine as far as I tested it.

Please take a look and share your feedback.

Thanks 🍻

drewnoakes commented:
@zvirja thanks for picking this up again.

I'm not totally sold on the use of the visitor pattern here. You mention it allows reuse, which is great, but a linq-style generator function would too and is a fair bit simpler. It's mostly a matter of taste but I find functional composition clearer than visitors. To add support for a new type, the whole hierarchy must be considered, and for someone coming into the code for the first time they have to navigate between multiple levels to understand the resulting behaviour. Subclasses also need to remember to call the base, or traversal breaks. I feel visitors are better suited to scenarios where there's a fixed set of cases to cover, such as XML attributes/elements/etc or AST nodes. The number of control subclasses is unbounded.

```c#
public static IEnumerable<Control> FindDescendants(this Control control)
{
    // ...
}

public static void AdjustForDpiScaling(this Control control)
{
    if (LicenseManager.UsageMode == LicenseUsageMode.Designtime)
    {
        return;
    }

    foreach (var descendant in control.FindDescendants()) 
    {
        switch (descendant)
        {
            case TextBox textBox:
                // ...
                break;
            // ...
        }
    }
}
```

We already have the first method:

https://github.com/gitextensions/gitextensions/blob/8f80b287cfcf6731161775861b7bbaeb2ea2bd06/GitUI/ControlUtil.cs#L9-L34

Note however that DPI scaling does not necessarily just apply to controls. At some point we will likely want to walk into items (such as tool strip items) which also have hierarchy, but a different base type. This could be done with a call to `IEnumerable<ToolStripItem> FindDescendants(this ToolStripItem item)` in a `case ToolStripItem toolStripItem` clause, without needing to introduce another visitor hierarchy.

This walking did originate for the purpose of DPI scaling and it looks like we want to use the walking operation for other purposes as well. It'd be good to do this with just a single traversal of the tree. With this change, we need to run two visitors. That's more CPU work and more memory allocation.

Interested to hear your thoughts.

zvirja commented:
Hi @drewnoakes, thanks for the review! 🙏

I definitely see your point and agree that indeed single `switch-case` block looks nicer than the visitor class with the inheritance. I'll try to rewrite using this approach. The only difference I can see is that with the `FindDescendants()` approach we _always_ traverse the `control.Children`, while currently we can decide whether we want to inspect children or skip them. Probably that doesn't actually matter, but formally that's a change.

As for the single traversal vs multi-traversal, I'd suggest to not mix things up and have two standalone methods, like `AdjustForDpiScaling()` and `EnableRemoveWordHotkeySupport()`. The cost isn't that much (tree is already in the memory and we don't allocate much), but it's much simpler to maintain the code by not mixing the responsibilities.

I'll rework my approach and get back to you when I finish 😉

drewnoakes commented on GitExtUtils/GitUI/ControlDpiExtensions.cs:
```diff
@@ -19,25 +18,19 @@ public static void AdjustForDpiScaling([NotNull] this Control control)
+                return;
+            }
 
+            foreach (var child in control.FindDescendants())
```
Nit: `child` should be `descendant`.

drewnoakes commented on GitExtUtils/GitUI/ControlHotkeyExtensions.cs:
```diff
@@ -0,0 +1,47 @@
+                return;
+            }
+
+            foreach (var child in control.FindDescendants())
```
Ditto.

drewnoakes commented on GitExtUtils/NativeMethods.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace GitExtUtils
+{
+    internal static class NativeMethods
```
Please make this a private inner class within `ControlHotkeyExtensions`.

zvirja commented:
@drewnoakes I applied the changes you asked about. No worries about the rework - I understand that you are worrying about the long-term project maintenance. Also I personally like how it looks now - more concise than before 😉 

As for the duplicate tree walk, probably we could create one more extension method like:
```c#
public static void WalkDescendants(this Control control, Action<Control> action) { ... }

// usage
this.WalkDescendants(AdjustForDpiScaling)
// or
this.WalkDescendants(d => { d.AdjustForDpiScaling(); d.EnableHotkey(); }
```

Of course, the `Adjust..` and `Enable..` extension methods should be rewritten to not perform the walk inside.

This approach doesn't sacrifice the maintenance and let you do the composition. The only drawback with this solution is that we wastefully traverse the tree in design mode and simply do nothing, but it's probably fine.

If you like this approach, I could adjust my code one more time 😉

vbjay commented:
How about a waker that raises an event for each control and all the
handlers do their jobs. You only walk the tree once and you can have as
many handlers as needed.

On Sat, Aug 4, 2018, 5:32 PM Alex Povar <notifications@github.com> wrote:

> @drewnoakes <https://github.com/drewnoakes> I applied the changes you
> asked about. No worries about the rework - I understand that you are
> worrying about the long-term project maintenance. Also I personally like
> how it looks now - more concise than before 😉
>
> As for the duplicate tree walk, probably we could create one more
> extension method like:
>
> public static void WalkDescendants(this Control control, Action<Control> action) { ... }
> // usagethis.WalkDescendants(AdjustForDpiScaling)// orthis.WalkDescendants(d => { d.AdjustForDpiScaling(); d.EnableHotkey(); }
>
> Of course, the Adjust.. and Enable.. extension methods should be
> rewritten to not perform the walk inside.
>
> This approach doesn't sacrifice the maintenance and let you do the
> composition. The only drawback with this solution is that we wastefully
> traverse the tree in design mode and simply do nothing, but it's probably
> fine.
>
> If you like this approach, I could adjust my code one more time 😉
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/5176#issuecomment-410478944>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADdhsSGcM_sGGSy_qMr2LxtiyXR5sxIrks5uNhL8gaJpZM4VUME6>
> .
>

zvirja commented:
@vbjay With this approach we are slowly progressing to the visitor pattern we came from 😅 

Well, I don't see a value event adds compared to the passed delegate. Moreover, it might be even more expensive to raise an event than to invoke a callback (you need to construct `EventArgs` at least...)

vbjay commented:
Then an iterator function and a for each. Still walk the tree but if you
turn it into an array you only walk the tree once.





On Sat, Aug 4, 2018, 5:40 PM Alex Povar <notifications@github.com> wrote:

> @vbjay <https://github.com/vbjay> With this approach we are slowly
> progressing to the visitor pattern we came from 😅
>
> Well, I don't see a value event adds compared to the passed delegate.
> Moreover, it might be even more expensive to raise an event than to invoke
> a callback (you need to construct EventArgs at least...)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/5176#issuecomment-410479265>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADdhsV5nv5mNAVDXPLWyclXLjYytdsdaks5uNhTBgaJpZM4VUME6>
> .
>

vbjay commented:
Of course turning the iterator function into an array would bump memory and
probably kill timing.

IEnumerable<Control> GetControls(Control ctrl)
{
Yield return  ctrl;
foreach (Control child in Ctrl.controls )
{
foreach( Control ct in GetControls(child)
{
Yield return ct;
}
}
}

I know it could be done without recursion but you get the idea.
On Sat, Aug 4, 2018, 5:42 PM Jay Asbury <vbjay.net@gmail.com> wrote:

> Then an iterator function and a for each. Still walk the tree but if you
> turn it into an array you only walk the tree once.
>
>
>
>
>
> On Sat, Aug 4, 2018, 5:40 PM Alex Povar <notifications@github.com> wrote:
>
>> @vbjay <https://github.com/vbjay> With this approach we are slowly
>> progressing to the visitor pattern we came from 😅
>>
>> Well, I don't see a value event adds compared to the passed delegate.
>> Moreover, it might be even more expensive to raise an event than to invoke
>> a callback (you need to construct EventArgs at least...)
>>
>> —
>> You are receiving this because you were mentioned.
>>
>>
>> Reply to this email directly, view it on GitHub
>> <https://github.com/gitextensions/gitextensions/pull/5176#issuecomment-410479265>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ADdhsV5nv5mNAVDXPLWyclXLjYytdsdaks5uNhTBgaJpZM4VUME6>
>> .
>>
>

RussKie commented:
I'm currently travelling, will review in next few days

On Sun, Aug 5, 2018, 7:57 AM Jay Asbury <notifications@github.com> wrote:

> Of course turning the iterator function into an array would bump memory and
> probably kill timing.
>
> IEnumerable<Control> GetControls(Control ctrl)
> {
> Yield return ctrl;
> foreach (Control child in Ctrl.controls )
> {
> foreach( Control ct in GetControls(child)
> {
> Yield return ct;
> }
> }
> }
>
> I know it could be done without recursion but you get the idea.
> On Sat, Aug 4, 2018, 5:42 PM Jay Asbury <vbjay.net@gmail.com> wrote:
>
> > Then an iterator function and a for each. Still walk the tree but if you
> > turn it into an array you only walk the tree once.
> >
> >
> >
> >
> >
> > On Sat, Aug 4, 2018, 5:40 PM Alex Povar <notifications@github.com>
> wrote:
> >
> >> @vbjay <https://github.com/vbjay> With this approach we are slowly
> >> progressing to the visitor pattern we came from 😅
> >>
> >> Well, I don't see a value event adds compared to the passed delegate.
> >> Moreover, it might be even more expensive to raise an event than to
> invoke
> >> a callback (you need to construct EventArgs at least...)
> >>
> >> —
> >> You are receiving this because you were mentioned.
> >>
> >>
> >> Reply to this email directly, view it on GitHub
> >> <
> https://github.com/gitextensions/gitextensions/pull/5176#issuecomment-410479265
> >,
> >> or mute the thread
> >> <
> https://github.com/notifications/unsubscribe-auth/ADdhsV5nv5mNAVDXPLWyclXLjYytdsdaks5uNhTBgaJpZM4VUME6
> >
> >> .
> >>
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/5176#issuecomment-410480064>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMyXvAuDsaq-EyMgoJrf7STnjnN3LCxks5uNhjAgaJpZM4VUME6>
> .
>

RussKie commented:
Thank you
